### PR TITLE
[dashboard] Allow workspace renaming

### LINKED ISSUE #3946: Bring back 'rename workspace' feature

### PR DESCRIPTION
**Pull request by laushinka**
## Description
After a discussion with @corneliusludmann and @gtsiolis, instead of the initial proposal of allowing workspace ID renames, this PR allows the renaming of a workspace context description. 
The reason to go this way is because the the workspaceID affects at the very least the URL of the workspace, and therefore could introduce breaking changes.

## Related Issue(s)
Fixes #3946

## How to test
1. Go to list of workspaces in Dashboard.
2. Click on the options on the far right, choose Rename.
3. Give the workspace a different description.

## Release Notes

```release-note
Allows renaming of workspace description.
```
Screenshot 2021-09-14 at 15 00 03
Screenshot 2021-09-14 at 15 00 17
Screenshot 2021-09-14 at 15 00 24

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#3946*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (gtsiolis) [inline components/dashboard/src/workspaces/WorkspaceEntry.tsx:161]: **suggestion**: Could we enforce a character limit let's say of 250 characters as a) it does not make sense to allow more characters since these won't be visible in the dashboard and b) the generic alert dialog when saving with a long description does not provide any useful context, right?

| Long description name hidden in the dashboard | Long description name error |
|-|-|
| Screenshot 2021-09-15 at 7 28 59 PM | Screenshot 2021-09-15 at 7 28 34 PM |

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (gtsiolis) [inline components/dashboard/src/workspaces/WorkspaceEntry.tsx:169]: **suggestion**: Using a more descriptive button label following the pattern used elsewhere sounds good. What do you think about this?

```suggestion
                <button className="ml-2" type="submit" onClick={updateWorkspaceDescription}>Update Description</button>
```

| Variables | Integrations | Workspaces |
|-|-|-|
| Screenshot 2021-09-15 at 7 36 04 PM | Screenshot 2021-09-15 at 7 36 29 PM | Screenshot 2021-09-15 at 7 36 42 PM |

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (gtsiolis) [inline components/dashboard/src/workspaces/WorkspaceEntry.tsx:161]: **issue**: Long descriptions seems to bring back a similar overflow issue as in https://github.com/gitpod-io/gitpod/issues/4556.

| Overflowing text when deleting a workspace |
|-|
| Screenshot 2021-09-15 at 7 36 55 PM (2) |

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (gtsiolis) [inline components/dashboard/src/workspaces/WorkspaceEntry.tsx:163]: **nitpick**: Re-read this and thought a better alternative could be something like the following. What do you think? This will overflow. 💡 

```suggestion
                    <p className="text-gray-500">Change the workspace description to make it easier to go back to a workspace.</p>
```

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (gtsiolis) [inline components/dashboard/src/workspaces/WorkspaceEntry.tsx:164]: **nitpick**: Minor typo.

```suggestion
                    <p className="text-gray-500">Workspace URLs and endpoints will remain the same.</p>
```

**Review by gtsiolis — APPROVED** [review]
UX look great overall, @laushinka! 🌟 

Left some comments below. 💭 

Approving to unblock merging. I'd suggest to also ask for a closer look (review) at the code. ➿ 

/hold

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (laushinka) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx:161]: Good catch!

Comment (gtsiolis) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx:161]: @laushinka Alternatively, we could not allow empty descriptions and return a warning alert. 💡

Comment (laushinka) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx:161]: We should truncate it I suppose?

Comment (gtsiolis) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx:161]: I'd vote for truncation to also avoid breaking this component!

On the other side, we could go all-in by a) wrapping the text, b) allowing longer description to surface on the dashboard, etc. _but_ this would be against the notion of ephemeral environments. Maybe this is something to consider in future if needed. 🗳️

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (gtsiolis) [inline components/dashboard/src/workspaces/WorkspaceEntry.tsx:131]: **thought**: Thinking of this PR and since this was a feature requested from some community members would it be interesting to measure how much people use this feature? Maybe have a self-serve graph of how many of the existing workspaces are renamed per day or week? Maybe also how many times people clicked the rename option in the dropdown option? It could be interesting to also see if the usage of this feature decreases after https://github.com/gitpod-io/gitpod/issues/3594. What do you think? Feel free to drop the idea if this would increase the scope of these changes.  💭

Cc @jakobhero

Comment (gtsiolis) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx:141]: **fyi**: Discussed this briefly with @svenefftinge earlier in the product sync meeting and we decided that adding tracking events here is a low priority given that the renaming functionality will be hidden behind the more actions dropdown. Up to you @laushinka and @jakobhero to decide if it's worth adding any tracking events within the scope of this PR. 🏓

Comment (laushinka) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx:143]: Agree that it would be useful to keep track but that we could do it in a separate issue if @jakobhero also thinks it's worth to track 🙏🏽

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (laushinka) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx]: Pushed a solution to show an error message and also truncating it. Let me know of a better error message! I'm off tomorrow, so feel free to change it here as well if I don't see the follow-up comment today.
Screenshot 2021-09-16 at 15 19 18

Comment (laushinka) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx]: Screenshot 2021-09-16 at 15 21 11

Comment (gtsiolis) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx]: **nitpick**: Nice! What do you think of using a softer tone like the following?

> Description cannot be empty.

**Comment by laushinka** [comment]
/unhold

@gtsiolis @corneliusludmann @JanKoehnlein just an fyi that I'll be off tomorrow.

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (laushinka) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx:186]: I committed the suggestion but somehow it still shows the old "Rename" button 🤔

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (laushinka) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx]: Yes 🍦

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5695#pullrequestreview-755266967" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/5695#issuecomment-920961200" title="LGTM">JanKoehnlein</a>*

Associated issue: *#3946*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [JanKoehnlein,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5695#pullrequestreview-755266967" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/5695#issuecomment-920975732" title="LGTM">JanKoehnlein</a>*

Associated issue: *#3946*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [JanKoehnlein,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/workspaces/WorkspaceEntry.tsx
Comment (jakobhero) [reply components/dashboard/src/workspaces/WorkspaceEntry.tsx:143]: hey there!👋 all clicks on buttons, anchors or clickable divs in the dashboard will be automatically tracked with the project & teams tracking PR i am currently working on. this means that we will see how many people clicked on the rename button, but now how many renames were successfully performed (which probably would be best tracked on the server). i think having visibility on how many users clicked on the button to open the modal is a good start to see how important this is to users and sufficient for now, wdyt?

**Pull request merged**
